WIP 3.1 Behaviors refactoring

robertpustulka opened the pull request:
*It's not ready for merge yet!*

Here is my attempt of moving file and image storage methods to dedicated behaviors.
Methods and callbacks from `FileStorageTable` has been moved to `FileStorageBehavior` and from `ImageStorageTable` to `ImageStorageBehavior` respectively.

I've removed the hardcoded checks for events which required the subject to be one of the `FileStorageTable` or `ImageStorageTable`. Instead listeners require events to have a table instance in it's data so storage events can be emitted from any class making them more flexible. Custom `EventDispatcherTrait` makes sure that a table object is present in events emitted by behaviors.

Old `ImageStorageTable` replaced `afterSave` and `afterDelete` callbacks from `FileStorageTable`, so `ImageStorageTable` didn't emit `FileStorage.afterSave` and `FileStorage.afterDelete` events. I emulated that behavior by detatching `FileStorageBehavior` listeners for those events inside `ImageStorageBehavior:initialize()`.

burzum commented:
> It's not ready for merge yet!

OK, I'll wait until it's done. Do you have any ETA on that?

robertpustulka commented:
>It should still work with old custom listeners

I haven't touched old listeners. They still should work the same. New listeners added in event overhaul branch are updated not to force event subject to be FileStorageTable or ImageStorageTable instances.

For BC events emitted by behaviors have subjects set to $this->_table by default.

robertpustulka commented:
Another thing is that new listeners don't support 'ImageVersion.getVersions' nor 'FileStorage.ImageHelper.imagePath' events, so version urls aren't retrieved in ImageHelper.

burzum commented:
I simply haven't had the time to implement them yet. Feel free to do so, if not I'll do it maybe next week,  some evening.

robertpustulka commented:
I've updated ImageVersionShell to use new events and added a few fixes. For example params check was done in initialize method which is called prior to options parsing. Now it's in startup.

burzum commented:
@robertpustulka do you have something left for this branch?

robertpustulka commented:
@burzum I don't have anything more to add. Just need to test it a bit more.

burzum commented:
Awesome! More testing would be good because the most I care about right now is not to screw up people who already use the old version and those who already *testing* (I hope they're aware of it) the overhaul branch. So BC **must** be 100%.

robertpustulka commented:
@burzum Prior to this change all operations for existing versions were just a waste - they did nothing because a target file had never been overwritten. 
I don't quite understand why you wouldn't want to overwrite versions you explicitly tell `createImageVersions()` to process. You tell which versions to update/create in `$versions`.

In shell you have 3 commands to manipulate versions:
```
generate    <model> <version> Generate a new image version
remove      <model> <version> Remove an new image version
regenerate  <model> Generates all image versions.
```

You can already tell which version you want to update. And also a way to update *all* versions.

robertpustulka commented:
I'm OK with adding an option to disable overwrite but as an if {} block for entire version process.

robertpustulka commented:
@burzum I think that people whose code depends on old listeners cannot use new listeners anyway so I'm not sure we should fight too much for BC there. The just have to upgrade their code.

I'm most concerned about this behavior refactoring. What if people override some methods in their tables and call `parent::someMethod()` inside them? Now these methods are moved to behaviors. I think I'm going to revert `ImageStorageTable` and `FileStorageTable` to their original form and encourage people to use behaviors after the upgrade to event-overhaul (3.1?).

robertpustulka commented:
@burzum OK, now there's a new switch in shell that allows for keeping old version files :)
By default all versions are overwritten

burzum commented:
@robertpustulka I guess you've already seen the failing test as well?

> Fatal error: Method name must be a string in /home/travis/build/burzum/cakephp-file-storage/src/Storage/Listener/ImageProcessingTrait.php on line 260

robertpustulka commented:
I have a favor to ask - could anyone test it on their setup? For me it seems to be working now, but I'd like somebody to test it against BC. I have no existing apps using old versions of the plugin.

robertpustulka commented:
@burzum I'm working on a fix right now.

burzum commented:
> have a favor to ask - could anyone test it on their setup? 

If you agree I'll merge this now into the 3.0 observer branch, I think some people already use it.

Again, not sure when I'll - or my company - has time for this any time soon. First day I'm writing code after a hell of planing and specification writing.

robertpustulka commented:
@burzum ok, let's merge it and test on a living organism :). It's still a dev branch, it could crash.

robertpustulka commented:
@burzum I haven't reverted the models. I'm still hesitating... I don't want to have the same code in two places. If people start to complain that this change breaks their tables then I'll try to find a good way to fix it.

burzum commented:
I *really* would like to keep it. There *is* any way duplicate code as long as the old listeners stay there. Check the roadmap in the wiki. I would like to provide a clear path instead of just breaking things out of a sudden.

robertpustulka commented:
@burzum ok, see #81
